src/utils/parse.ts

```diff
@@ -0,0 +1,74 @@
+import Decimal from 'decimal.js';
+import ExchangeRate from '../models/ExchangeRate';
+import { ExchangeRateJSON, CurrencyCode, Dictionary, TimeInterval, Network } from '../types';
+
+/** @internal */
+const parseFiatMap = (fiatMapJSON: Record<string, string>) => {
+  const fiatMap: Dictionary<CurrencyCode, Decimal> = {};
+  Object.keys(fiatMapJSON).forEach((currencyCode) => {
+    fiatMap[currencyCode as CurrencyCode] = new Decimal(fiatMapJSON[currencyCode]);
+  });
+  return fiatMap;
+};
+
+/** @internal */
+const parseExchangeAddressMap = (exchangeAddressMapJSON: Record<string, string>) => {
+  const exchangeAddressMap: Dictionary<Network, string> = {};
+  Object.keys(exchangeAddressMapJSON).forEach((network) => {
+    exchangeAddressMap[network as Network] = exchangeAddressMapJSON[network];
+  });
+  return exchangeAddressMap;
+};
+
+/** @internal */
+const parseExchangeRates = (
+  exchangeRateMapJSON: Record<string, Record<string, ExchangeRateJSON>>
+) => {
+  const exchangeRatesMap: Dictionary<CurrencyCode, Dictionary<CurrencyCode, ExchangeRate>> = {};
+  Object.keys(exchangeRateMapJSON).forEach((depositCurrency) => {
+    const innerMap: Dictionary<CurrencyCode, ExchangeRate> = {};
+    Object.keys(exchangeRateMapJSON[depositCurrency]).forEach((toCurrency) => {
+      innerMap[toCurrency as CurrencyCode] = new ExchangeRate(
+        exchangeRateMapJSON[depositCurrency][toCurrency]
+      );
+    });
+    exchangeRatesMap[depositCurrency as CurrencyCode] = innerMap;
+  });
+  return exchangeRatesMap;
+};
+
+const parseHistoricalExchangeRates = (
```

Comment:
This is one nice function 🚀

Reply:
This function is the story of my life. There are there more of them this in Android, Objective-C and C+:

```java
public static WritableMap mapHistoricalExchangeRates(HashMap<String, HashMap<String, HashMap<String, ArrayList<ExchangeRate>>>> historicalExchangeRates) {

    WritableMap outerOuterMap = Arguments.createMap();

    for (HashMap.Entry<String, HashMap<String, HashMap<String, ArrayList<ExchangeRate>>>> outerOuterEntry : historicalExchangeRates.entrySet()) {

      String timeInterval = outerOuterEntry.getKey();
      WritableMap outerMap = Arguments.createMap();

      for (HashMap.Entry<String, HashMap<String, ArrayList<ExchangeRate>>> outerEntry : outerOuterEntry.getValue().entrySet()) {

        String fromCurrency = outerEntry.getKey();
        WritableMap innerMap = Arguments.createMap();

        for (HashMap.Entry<String, ArrayList<ExchangeRate>> innerEntry : outerEntry.getValue().entrySet()) {

          String toCurrency = innerEntry.getKey();

          WritableArray array = Arguments.createArray();

          for (ExchangeRate rate : innerEntry.getValue()) {
            WritableMap map = RNZumoKitModule.mapExchangeRate(rate);
            array.pushMap(map);
          }

          innerMap.putArray(toCurrency, array);
        }

        outerMap.putMap(fromCurrency, innerMap);
      }

      outerOuterMap.putMap(timeInterval, outerMap);
    }

    return outerOuterMap;

  }
```

```objc
- (NSDictionary *)mapHistoricalExchangeRates:(ZKHistoricalExchangeRates) historicalRates {

     NSMutableDictionary *rates = [[NSMutableDictionary alloc] init];

     [historicalRates enumerateKeysAndObjectsUsingBlock:^(
             NSString * _Nonnull timeInterval,
             NSDictionary<NSString *, NSDictionary<NSString *, NSArray<ZKExchangeRate *> *> *> * _Nonnull exchangeRates,
             BOOL * _Nonnull stop) {

         NSMutableDictionary *outerDict = [[NSMutableDictionary alloc] init];

         [exchangeRates enumerateKeysAndObjectsUsingBlock:^(NSString * _Nonnull outerKey, NSDictionary<NSString *, NSArray<ZKExchangeRate *> *> * _Nonnull outerObj, BOOL * _Nonnull stop) {

             NSMutableDictionary *innerDict = [[NSMutableDictionary alloc] init];

             [outerObj enumerateKeysAndObjectsUsingBlock:^(NSString * _Nonnull innerKey, NSArray<ZKExchangeRate *> * _Nonnull array, BOOL * _Nonnull stop) {

                NSMutableArray<ZKExchangeRate *> *mapped = [[NSMutableArray alloc] init];

                [array enumerateObjectsUsingBlock:^(
                        ZKExchangeRate * _Nonnull obj,
                        NSUInteger idx,
                        BOOL * _Nonnull stop) {

                    [mapped addObject:[self mapExchangeRate:obj]];

                }];

                 innerDict[innerKey] = mapped;
             }];

             outerDict[outerKey] = innerDict;
         }];

         rates[timeInterval] = outerDict;
     }];

     return rates;
 }
```

```cpp
static json historical_exchage_rates(json j)
{
    json res;
    for (auto &item : j.items()) {
        string interval = item.key();
        res[interval] = json({});

        for (auto &exchange_rates_map_json : item.value().items()) {
            auto exchange_rates_map_patched = exchange_rates_map(exchange_rates_map_json.value());

            for (const auto &outer : exchange_rates_map_patched.items()) {
                string from_currency = outer.key();
                if (res.at(interval).find(from_currency) == res.at(interval).end())
                    res[interval][from_currency] = json({});

                for (const auto &inner : outer.value().items()) {
                    string to_currency = inner.key();
                    if (res.at(interval).at(from_currency).find(to_currency) ==
                        res.at(interval).at(from_currency).end())
                        res[interval][from_currency][to_currency] = json::array();

                    res.at(interval).at(from_currency).at(to_currency).push_back(inner.value());
                }
            }
        }
    }

    return res;
}
```